src/graph/visitor/ExtractGroupSuiteVisitor.cpp

```diff
@@ -0,0 +1,203 @@
+  pushGroupSuite(expr);
+}
+
+void ExtractGroupSuiteVisitor::visit(EdgePropertyExpression *expr) {
+  pushGroupSuite(expr);
+}
+
+void ExtractGroupSuiteVisitor::visit(InputPropertyExpression *expr) {
+  pushGroupSuite(expr);
+}
+
+void ExtractGroupSuiteVisitor::visit(VariablePropertyExpression *expr) {
+  pushGroupSuite(expr);
+}
+
+void ExtractGroupSuiteVisitor::visit(DestPropertyExpression *expr) {
+  pushGroupSuite(expr);
+}
+
+void ExtractGroupSuiteVisitor::visit(SourcePropertyExpression *expr) {
+  pushGroupSuite(expr);
+}
+
+void ExtractGroupSuiteVisitor::visit(EdgeSrcIdExpression *expr) {
+  pushGroupSuite(expr);
+}
+
+void ExtractGroupSuiteVisitor::visit(EdgeTypeExpression *expr) {
+  pushGroupSuite(expr);
+}
+
+void ExtractGroupSuiteVisitor::visit(EdgeRankExpression *expr) {
+  pushGroupSuite(expr);
+}
+
+void ExtractGroupSuiteVisitor::visit(EdgeDstIdExpression *expr) {
+  pushGroupSuite(expr);
+}
+
+void ExtractGroupSuiteVisitor::visit(VertexExpression *expr) {
+  pushGroupSuite(expr);
+}
+
+void ExtractGroupSuiteVisitor::visit(EdgeExpression *expr) {
+  pushGroupSuite(expr);
+}
+
+void ExtractGroupSuiteVisitor::visit(CaseExpression *expr) {
+  internalVisit(expr);
+}
+
+void ExtractGroupSuiteVisitor::visit(PathBuildExpression *expr) {
+  internalVisit(expr);
+}
+
+void ExtractGroupSuiteVisitor::visit(ColumnExpression *expr) {
+  pushGroupSuite(expr);
+}
+
+void ExtractGroupSuiteVisitor::visit(PredicateExpression *expr) {
+  internalVisit(expr);
+}
+
+void ExtractGroupSuiteVisitor::visit(ListComprehensionExpression *expr) {
+  internalVisit(expr);
+}
+
+void ExtractGroupSuiteVisitor::visit(ReduceExpression *expr) {
+  internalVisit(expr);
+}
+
+void ExtractGroupSuiteVisitor::visit(SubscriptRangeExpression *expr) {
+  internalVisit(expr);
+}
+
+void ExtractGroupSuiteVisitor::visit(MatchPathPatternExpression *expr) {
+  internalVisit(expr);
+}
+
+template <typename T>
+void ExtractGroupSuiteVisitor::internalVisit(T *expr) {
+  if (ExpressionUtils::hasAny(expr, {Expression::Kind::kAggregate})) {
+    ExprVisitorImpl::visit(expr);
+  } else {
+    pushGroupSuite(expr);
+  }
+}
+
+template <typename T>
+void ExtractGroupSuiteVisitor::pushGroupSuite(T *expr) {
+  // When expr is PredicateExpression, ListComprehensionExpression or ReduceExpression,
+  // it needs to check whether contains innerVar. If so, it doesn't need to push groupKeys and
+  // groupItems.
+  auto specialExprs =
+      ExpressionUtils::collectAll(expr, {Expression::Kind::kVar, Expression::Kind::kAttribute});
+  for (auto *s : specialExprs) {
+    if (s->kind() == Expression::Kind::kVar &&
+        static_cast<const VariableExpression *>(s)->isInner()) {
+      return;
+    }
+    if (s->kind() == Expression::Kind::kAttribute) {
+      auto *left = static_cast<const AttributeExpression *>(s)->left();
+      if (left->kind() == Expression::Kind::kVar &&
+          static_cast<const VariableExpression *>(left)->isInner()) {
+        return;
+      }
```

Comment:
Why need this?

Reply:
It is for PredicateExpression, ListComprehensionExpression or ReduceExpression. It will check the innerVar.

Reply:
Why collect AttributeExpression, but still check VariableExpression?
Is this code redundant?